core/src/main/java/io/lionweb/lioncore/java/language/LionCoreBuiltins.java

```diff
@@ -38,30 +47,61 @@ private LionCoreBuiltins() {
   }
 
   public static LionCoreBuiltins getInstance() {
-    return INSTANCE;
+    return getInstance(LionWebVersion.currentVersion);
   }
 
   public static PrimitiveType getString() {
-    return INSTANCE.getPrimitiveTypeByName("String");
+    return getInstance().getPrimitiveTypeByName("String");
   }
 
   public static PrimitiveType getInteger() {
-    return INSTANCE.getPrimitiveTypeByName("Integer");
+    return getInstance().getPrimitiveTypeByName("Integer");
   }
 
   public static PrimitiveType getBoolean() {
-    return INSTANCE.getPrimitiveTypeByName("Boolean");
-  }
-
-  public static PrimitiveType getJSON() {
-    return INSTANCE.getPrimitiveTypeByName("JSON");
+    return getInstance().getPrimitiveTypeByName("Boolean");
   }
 
   public static Interface getINamed() {
-    return INSTANCE.getInterfaceByName("INamed");
+    return getInstance().getInterfaceByName("INamed");
   }
 
   public static Concept getNode() {
-    return INSTANCE.getConceptByName("Node");
+    return getInstance().getConceptByName("Node");
+  }
+
+  public static LionCoreBuiltins getInstance(@Nonnull LionWebVersion lionWebVersion) {
```

Comment:
I think it will be close to impossible to control that all users of this method will use it with the correct version, so mixing them will be pretty common.

Reply:
Yes, that could be the case. However I would expect these methods to be used in "infrastructure" code (like the repository, libraries to implement migrations, etc.). Everyday users would probably not use these methods frequently. Also, I think that most users would just use the variants not specifying a version because they may want to always use the latest one, unless they specifically need to deal with "legacy" models

Reply:
I'm currently trying in the C# version whether I can use the compiler to guarantee this (would also dictate the [issue above](https://github.com/LionWeb-io/lionweb-java/pull/167#discussion_r1847186860)). I'll let you know how I fare.

Reply:
I was not successful in C#. However, my guess is it would work in Java. It essentially failed because I cannot write `SomeType<?>` in C# -- but Java allows it.

Reply:
This was my idea:

I have a marker interface `IVersionDependent` that's implemented by every class that has to work differently for different versions. It has a generic type parameter `T extends IVersion` to bind it to some version.
Each version is represented by a class that implements `IVersion`.

Our guinea pig is a `Serializer`. It has to process builtin primitives differently depending on the version.
Thus we define an interface `IProcessBuiltinPrimitives`, and one implementation for each version.

When we create a serializer for a specific version, we have to create the correct variant for processing builtin primitives.
This is the part that doesn't work in C#, but should work in Java.

Not shown here: All M2 languages also implement `IVersionDependent`, so the compiler makes sure we only pass languages to the serializer with the correct version: `Serializer<TSerializerVersion> { void RegisterLanguage<TLangVersion>(Language<TLangVersion> language) where TLangVersion : TSerializerVersion {} }`

```csharp

interface IVersion;

class V1 : IVersion;
class V2 : IVersion;

interface IVersionDependent<T> where T : IVersion;

class Serializer<T> : IVersionDependent<T> where T : IVersion
{
    public string Name { get; } = "me!";
    public IDelegate<T> Deleg { get; init; }

    public void Run() => Deleg.DoIt(this);
}

interface IProcessBuiltinPrimitives<T> where T : IVersion
{
    public void DoIt(Serializer<T> serial);
}

abstract class ProcessBuiltinPrimitives<T> : IProcessBuiltinPrimitives<T> where T : IVersion
{
    public abstract void DoIt(Serializer<T> serial);
}

class ProcessBuiltinPrimitives_V1 : ProcessBuiltinPrimitives<V1>
{
    public override void DoIt(Serializer<T> serial) => Console.WriteLine($"V1{serial.Name}");
}

class ProcessBuiltinPrimitives_V2: ProcessBuiltinPrimitives<V2>
{
    public override void DoIt(Serializer<T> serial) => Console.WriteLine($"V2{serial.Name}");
}

class SerializerBuilder<Tx> : IVersionDependent<Tx> where Tx : IVersion
{
    public Serializer<T> CreateSerializer<T>(T version) where T : Tx
    {
        // Would need IProcessBuiltinPrimitives<?>
        IProcessBuiltinPrimitives<T> deleg = version switch
        {
            V1 => new ProcessBuiltinPrimitives_V1(),
            V2 => new ProcessBuiltinPrimitives_V2()
        };

        var serial = new Serializer<T>() { Deleg = deleg };
        return serial;
    }
}
```

Reply:
It does work:

```java
public class Bla {
    interface IVersion{}

    static class V1 implements IVersion{
        public String toString() { return "V1"; }
    }
    static class V2 implements IVersion{
        public String toString() { return "V2"; }
    }

    interface IVersionDependent<T extends IVersion> {}

    static class Language<T extends IVersion> implements IVersionDependent<T> {
        private final T version;
        private final String name;

        public Language(T version, String name) {
            this.version = version;
            this.name = name;
        }

        public String toString() { return "Language{" + "version=" + version + ", name='" + name + '\'' + '}'; }
    }

    static class Serializer<T extends IVersion> implements IVersionDependent<T>
    {
        private final IProcessBuiltinPrimitives<T> deleg;
        public Serializer(IProcessBuiltinPrimitives<T> deleg) { this.deleg = deleg; }

        public String getName() { return "me"; }
        public String Run(Language<T> language) {return deleg.DoIt(this) + " " + language;}
    }

    interface IProcessBuiltinPrimitives<T extends IVersion>
    {
        String DoIt(Serializer<T> serial);
    }

    static class ProcessBuiltinPrimitives_V1 implements IProcessBuiltinPrimitives<V1>
    {
        public String DoIt(Serializer<V1> serial) { return "V1" + serial.getName(); }
    }

    static class ProcessBuiltinPrimitives_V2 implements IProcessBuiltinPrimitives<V2>
    {
        public String DoIt(Serializer<V2> serial) { return "V2" + serial.getName(); }
    }

    static class SerializerBuilder<T extends IVersion> implements IVersionDependent<T>
    {
        private final T version;

        public SerializerBuilder(T version) { this.version = version; }

        public Serializer<T> CreateSerializer()
        {
            IProcessBuiltinPrimitives<T> deleg = (IProcessBuiltinPrimitives<T>) (version instanceof V1 ?  new ProcessBuiltinPrimitives_V1() :  new ProcessBuiltinPrimitives_V2());

            return new Serializer<>(deleg);
        }
    }

    @Test
    public void test() {
        Language<V1> v1Lang = new Language<>(new V1(), "v1Lang");
        Language<V2> v2Lang = new Language<>(new V2(), "v2Lang");

        Assert.assertEquals("V1me Language{version=V1, name='v1Lang'}", new SerializerBuilder<>(new V1()).CreateSerializer().Run(v1Lang));
        // Doesn't compile
        // Assert.assertEquals("V2me Language{version=V1, name='v1Lang'}", new SerializerBuilder<V2>(new V2()).CreateSerializer().Run(v1Lang));
        Assert.assertEquals("V2me Language{version=V2, name='v2Lang'}", new SerializerBuilder<>(new V2()).CreateSerializer().Run(v2Lang));
    }
}
```

Reply:
I thought about using generics, but I was afraid it would make the code more cumbersome to use. Forcing users to write more complex code and having casts here and there.

Admittedly I did not think about using a token type for versions as you did.

Perhaps we can do an experiment with this idea to see how complex the code gets and then decide if the safety it provides is worth the complexity. Would this make sense?

Reply:
I started the experiment. I am writing a lot of `<` and `>` :D 
I will share it as soon as I can